contracts/core/Allo.sol

```diff
@@ -1,79 +1,359 @@
+    /// ====================================
+
+    /// @notice Initializes the contract after an upgrade
+    /// @dev During upgrade -> an higher version should be passed to reinitializer
+    /// @param _registry The address of the registry
+    /// @param _treasury The address of the treasury
+    /// @param _feePercentage The fee percentage
+    function initialize(address _registry, address payable _treasury, uint256 _feePercentage) public reinitializer(1) {
+        _initializeOwner(msg.sender);
+
+        registry = Registry(_registry);
+        treasury = _treasury;
+        feePercentage = _feePercentage;
+
+        emit RegistryUpdated(_registry);
+        emit TreasuryUpdated(_treasury);
+        emit FeeUpdated(_feePercentage);
+
+        __Multicall_init();
+    }
+
+    /// ====================================
+    /// =========== Modifier ===============
+    /// ====================================
+
+    modifier isPoolAdmin(uint256 _poolId) {
+        if (!registry.isOwnerOrMemberOfIdentity(pools[_poolId].identityId, msg.sender)) {
+            revert NO_ACCESS_TO_ROLE();
+        }
+        _;
+    }
+
+    /// ====================================
+    /// ==== External/Public Functions =====
+    /// ====================================
+
+    /// @notice Returns the pool info
+    /// @param _poolId The id of the pool
+    /// @dev calls out to the registry to get the identity metadata
+    function getPoolInfo(uint256 _poolId) external view returns (Pool memory) {
+        return pools[_poolId];
+    }
+
+    /// @notice Creates a new pool (with clone)
+    /// @param _identityId The identityId of the pool
+    /// @param _allocationStrategy The address of the allocation strategy
+    /// @param _distributionStrategy The address of the distribution strategy
+    /// @param _token The address of the token
+    /// @param _amount The amount of the token
+    /// @param _metadata The metadata of the pool
+    /// @param _cloneAllocationStrategy Boolean flag to clone the allocation strategy
+    /// @param _cloneDistributionStrategy Boolean flag to clone the distribution strategy
+    function createPoolWithClone(
+        bytes32 _identityId,
+        address _allocationStrategy,
+        address payable _distributionStrategy,
+        address _token,
+        uint256 _amount,
+        Metadata memory _metadata,
+        bool _cloneAllocationStrategy,
+        bool _cloneDistributionStrategy
+    ) external payable returns (uint256 poolId) {
+        /// Check the boolean flag to see if the strategy should be cloned
+        /// Additionally, force clone the strategy if it's approved
+        address allocationStrategy =
+            _cloneAllocationStrategy ? Clone.createClone(_allocationStrategy, _nonce++) : _allocationStrategy;
+
+        address distributionStrategy =
+            _cloneDistributionStrategy ? Clone.createClone(_distributionStrategy, _nonce++) : _distributionStrategy;
```

Comment:
Do we want to check `_isApprovedStrategy()` here? I think this would be the best place.

Reply:
Do we want to do something like this?

```javascript
if (!_isApprovedStrategy(_allocationStrategy) && !_isApprovedStrategy(_distributionStrategy)) {
    revert NOT_APPROVED_STRATEGY();
}

address allocationStrategy =
      _cloneAllocationStrategy ? Clone.createClone(_allocationStrategy, _nonce++) : _allocationStrategy;

address distributionStrategy =
  _cloneDistributionStrategy ? Clone.createClone(_distributionStrategy, _nonce++) : _distributionStrategy;
```
or
```javascript
address allocationStrategy =
    _cloneAllocationStrategy && _isApprovedStrategy(_allocationStrategy)
    ? Clone.createClone(_allocationStrategy, _nonce++)
    : _allocationStrategy;

address distributionStrategy = _cloneDistributionStrategy && _isApprovedStrategy(_distributionStrategy)
    ? Clone.createClone(_distributionStrategy, _nonce++)
    : _distributionStrategy;
```

What about if they wanted to create a clone of their own, unapproved contract?

Reply:
They can't create a clone of their own unapproved contract on Allo.sol. They'd need to do that themselves. 

Clones on this contract exist for cloning our approved contracts, nothing else!

Reply:
It would be more like:
```
if (_cloneAllocationStrategy) {
    require(_isApprovedStrategy(_allocationStrategy));
    ...
} 

if (_cloneDistributionStrategy) {
    require(...)
}
```

Reply:
Should we use `require()` or custom error?

```javascript
if (!_isApprovedStrategy(_allocationStrategy)) {
    revert STRATEGY_NOT_APPROVED();
}
```

`require` is a simpler and more straightforward approach, especially when the error message doesn't require detailed information. Custom error handling, on the other hand, allows us to provide more context-specific error messages, which can be helpful for debugging/testing and providing meaningful feedback to our users.

What are your thoughts?

Reply:
I'm just giving you pseudocode of the idea of where checks should happen. Stay consistent across the contract with custom errors for sure.

Reply:
Understood, thank you 🙏

Reply:
> They can't create a clone of their own unapproved contract on Allo.sol. They'd need to do that themselves.

> It would be more like:
> 
> ```
> if (_cloneAllocationStrategy) {
>     require(_isApprovedStrategy(_allocationStrategy));
>     ...
> } 
> 
> if (_cloneDistributionStrategy) {
>     require(...)
> }
> ```

This might not work cause if we would want to allow a user to clone an approved allocation strategy but use their custom distribution strategy.

Given the requirements: 
- We want to allow cloning of ONLY approved contracts
- We want to allow people to one approved strategy (clone) + one custom strategy 

Here is a simpler way we can do this ad2c6d6

```javascript
    function createPoolWithClone(
        bytes32 _identityId,
        address _allocationStrategy,
        address payable _distributionStrategy,
        address _token,
        uint256 _amount,
        Metadata memory _metadata
    ) external payable returns (uint256 poolId) {
        address allocationStrategy =
            _isApprovedStrategy(_allocationStrategy) ? Clone.createClone(_allocationStrategy) : _allocationStrategy;
        address distributionStrategy = _isApprovedStrategy(_distributionStrategy)
            ? Clone.createClone(_distributionStrategy)
            : _distributionStrategy;

        return _createPool(_identityId, allocationStrategy, payable(distributionStrategy), _token, _amount, _metadata);
    }
    ```

Reply:
Yes, I agree. We had that in there this morning and adding it back now.